### ISSUE TITLE
[Experimental idea] Redesign section naming

### ISSUE THREAD

**Issue by SimonBiggs** [SimonBiggs/scriptedforms#247]
Some section names are don't currently clearly highlight what they actually achieve. I propose changing section names in the following way:

* `<section-button>` --> `<code-trigger-button>`
* `<section-live>` --> `<code-trigger-content>`
* `<section-output>` --> `<code-trigger-global>`
* `<section-filechange>` --> `<code-trigger-filechange>`
* `<section-start>` --> `<code-trigger-start>`

I also propose creating a more advanced code trigger component which can be customised to be mixes of the above as required:

```html
<code-trigger triggers="a_python_list" options="a_python_dict">
```

**Comment by SimonBiggs** [comment]
@robmarkcole along the same "keeping it simple" vein, does this proposal keep the basics simple do you think? Should I be looking at section naming redesign, or is it okay as is?

**Comment by robmarkcole** [comment]
I like the existing namings, but that’s just a personal preference as they are a bit more succinct

**Comment by SimonBiggs** [comment]
@robmarkcole small iteration on the proposal, what do you think about that?

**Comment by SimonBiggs** [comment]
I don't believe this is name changing for name changing sake. At the beginning I only had start, button and live. But now that I have 'output' and 'filechange' as well as live the name 'live' doesn't make much sense anymore. It was chosen to be live in contrast to 'button'. But 'filechange' is arguably also 'live', so is 'output'.

So I believe I need to change live to 'contents' or something of the sort making it clear that live only runs when a variable within it changes.

But I have also often come across cases where I want a section to do two things. So I do also need to build in the ability to customise the sections. Given that I am already changing the name of 'live' it means that the choice of the word 'section' should also be considered to make sure it is the best option.

**Comment by SimonBiggs** [comment]
To play devils advocate the "section" helps indicate where that element
needs to be placed. It helps indicate that it needs to go around the code
blocks. "Trigger" doesn't give you a an idea about it needing to surround
the cod blocks.

The live name can change to content or something similar. But keep
"section" as is?

**Comment by SimonBiggs** [comment]
@robmarkcole 

Here's another thought. Go back to just `<section-start>`, `<section-button>`, and `<section-live>`. Remove filechange and output. And instead have a trigger parameter on `<section-live>`.

The trigger parameter on the live section would default to `trigger="['content']"`, but could be replaced with options such as `trigger="['global']"` or `trigger="['file']"`

**Comment by SimonBiggs** [comment]
or what if the parameter was instead 'addTriggers' which indicated that it would always keep 'content' but could have others included. And make it so that the 'addTriggers' parameter is available on all sections.

This would mean that `<section-start addTriggers="['content']">` would actually be equivalent to `<section-live addTriggers="['start']">`

**Comment by SimonBiggs** [comment]
Hmm, different words altogether. Inspired by RxJs (http://reactivex.io/documentation/operators/subscribe.html)

What about:

* `<observe-files>`
* `<observe-all-variables>`
* `<observe-content-variables>`
* `<observe-code-result>`
* `<observe-session-start>`
* `<observe-page-load>`
* `<observe-button>`

* `<observe-custom subscriptions="['files', 'button']">`

This is especially poignant because that is exactly what the underlying code is doing. It is using the RxJs library to observe changes in those respective things.

**Comment by robmarkcole** [comment]
@SimonBiggs I like that formalism

**Comment by SimonBiggs** [comment]
Woo hoo! :) we have a winner!

Okay, back up a sec, why do you like it? (I don't want to race ahead with the idea unless we're sure that we're sure).

**Comment by SimonBiggs** [comment]
I'm just going to iterate ideas about what the parameters should look like in this comment.

`<observe-files files="['a_file_name']">`

`<observe-all-variables>`

`<observe-content-variables>`

`<observe-code-result codeResult="update_when_this_function_result_changes()">`

`<observe-session-start>`

`<observe-page-load>`

`<observe-button value="a_button_name">`

`<observe-custom subscriptions="['files', 'button']" options="{'files': ['a_file_name'], 'button': {'value': 'a_button_name'}}">`

**Comment by SimonBiggs** [comment]
I think I have managed to convince myself. Managing to convince you @robmarkcole so quickly probably helped sell myself on the idea.

**Comment by SimonBiggs** [comment]
So, unless there are any objections I will begin implementing this formalism in parallel. The old method will hang around. Once the new method is complete the old method will get deprecation notices.

I will keep the old method around for quite a while though to make the transition smooth.

The new interactive docs will only make mention from the get go this new syntax.

**Comment by SimonBiggs** [comment]
@robmarkcole do you have any thoughts about the parameter design?

**Comment by robmarkcole** [comment]
@SimonBiggs could you clarify with a single example to choose between?

**Comment by SimonBiggs** [comment]
For example:

* changing conditional to disabled, resulting in being more aligned to standard html
* In observe code result, trying to make it clear which is the code being run and which code is being watched
* Change the parameter code parameter from code to runCode, primarily to facilitate the point above
* If options are wanted on the observe custom, pass a dictionary. The keys of that dictionary will be used to work out what subscriptions are needed.

**Comment by robmarkcole** [comment]
@SimonBiggs 
- Consistency with standard HTML recommended
- Clarifying watched vs. run code is necessary
- runCode is useful
- Dict is a good approach

**Comment by SimonBiggs** [comment]
Thanks Robin. Let me know if I've been sending too many pings your way.

**Comment by SimonBiggs** [comment]
@robmarkcole another question, could I get away with:

~~~html
<observe-code-result code="a_function()" runCode="some_code()"> </observe-code-result>
~~~

As in if code to be run is always called "runCode" is it obvious enough to use "code" instead of "observe code" on the observe-code-result section?

**Comment by SimonBiggs** [comment]
@robmarkcole and please let me know if you can think of ways to shorten any of the names and still make it clear what the mean.

**Comment by robmarkcole** [comment]
Hi @SimonBiggs I think if you are using `runCode` then you should also use `observeCode` : explicit is better than implicit :)

**Comment by SimonBiggs** [comment]
Kk :) sounds good. Explicit is definitely better than implicit. Thanks
Robin.

**Comment by SimonBiggs** [comment]
@robmarkcole here's a project that is very very similar in syntax. It's designed to produce static documents, but the syntax has been well thought out and the same syntax could achieve the same ends here.

https://rmarkdown.rstudio.com/lesson-3.html

Essentially don't have sections at all. Instead define how each code block interacts in the header of the code block.

~~~markdown
```{python button}
print('hello world')
```
~~~

I'm not sold on the idea. But thought the idea needed to be raised seeing that r markdown is achieving a very similar aim with a very similar syntax.

**Comment by robmarkcole** [comment]
@simonbiggs as someone with no web dev experience but plenty of python experience, that syntax does have a ‘familiar look and feel’. I’ve been working out how to best implement some syntax in my own project recently and the tricky part is establishing the best structural relationships between syntax. If that package has those relationships figured out already you could just adopt their structure but change the precise syntax.

**Comment by SimonBiggs** [comment]
Interesting. I'll think more about it. I'm not really sold on it...

If you could think of some ways the parameters could be neatly put in there
that would help. I can't picture how to get a similar level of
flexibility...

**Comment by robmarkcole** [comment]
For parameter entry could you use a python structure such as a dict?

**Comment by SimonBiggs** [comment]
Would you be able to write out what you think it might look like?

Say for a button section, and a filechange section.
